### PR TITLE
Add firing of events when migration and rollback happen

### PR DESCRIPTION
**Pull request by sane4ek-2**
Sometimes we want do some things regularly after migrations or rollbacks. Before that changes we couldn't do that. But now we can make our own event subscribers and they will be called when one of the two commands call. Also we can refuse from subscribers and don't write them.

### CODE REVIEW

File: src/Phinx/Console/CommandEvents.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Phinx\Console;
+
+class CommandEvents
```
Comment (rquadling) [inline src/Phinx/Console/CommandEvents.php]: I can certainly see use cases for phinx.command.[migrate|rollback|seed|breakpoint].[started|completed|failed]

Well. Maybe not so much breakpoint, but keep the pattern as then you don't have "special" behaviour.

Also, (maybe overkill), would firing an event for each migration/seed be possible?

phinx.command.[migrate-single|rollback-single|seed-single].[started|completed|failed]

The event data for the singles (not a great name admittedly, so be inventive!!!) would need to include the migration name.

One use case for this for me would be that I could activate an "after" migration once a "before" migration has been executed.

Comment (sane4ek-2) [reply src/Phinx/Console/CommandEvents.php]: > I can certainly see use cases for phinx.command.[migrate|rollback|seed|breakpoint].[started|completed|failed]

I do that.

> Also, (maybe overkill), would firing an event for each migration/seed be possible?
> phinx.command.[migrate-single|rollback-single|seed-single].[started|completed|failed]
> The event data for the singles (not a great name admittedly, so be inventive!!!) would need to include the migration name.

If be honest I don't see any use case right now. I suggest not to develop that at present time.

> One use case for this for me would be that I could activate an "after" migration once a "before" migration has been executed.

You're right. Let's do it.

File: src/Phinx/Console/Command/AbstractCommand.php [hunk 2]
```diff
@@ -362,4 +374,31 @@ protected function getSeedTemplateFilename()
...
+        }
+        $finder = new Finder();
+        $listeners = $finder->in($paths)->name('*.php')->files();
+
+        /** @var SplFileInfo $listener */
+        foreach ($listeners as $listener) {
+            require_once $listener->getRealPath();
+
+            $refClass = new \ReflectionClass($listener->getBasename('.php'));
+            if (!$refClass->implementsInterface('\Symfony\Component\EventDispatcher\EventSubscriberInterface')) {
```
Comment (dereuromark) [inline src/Phinx/Console/Command/AbstractCommand.php]: Would be better to use ::class usage here.

File: src/Phinx/Console/Command/AbstractCommand.php [hunk 3]
```diff
@@ -362,4 +374,31 @@ protected function getSeedTemplateFilename()
...
+    protected function loadEventSubscribers()
+    {
+        $paths = $this->getConfig()->getSubscriberPaths();
+        if (!$paths) {
+            return;
+        }
+        $finder = new Finder();
+        $listeners = $finder->in($paths)->name('*.php')->files();
+
+        /** @var SplFileInfo $listener */
```
Comment (dereuromark) [inline src/Phinx/Console/Command/AbstractCommand.php]: FQCN missing

**Comment by dersam** [comment]
I'd love to be able to define events around migrations - I have a few use cases that would be solved by an event system. What still needs to be done to get this to a point where it would be mergeable? I'd be willing to finish it off if @sane4ek-2 isn't free to complete it.

**Comment by sane4ek-2** [comment]
@dersam it would be great if you could complete those changes. I forget about this PR and don't have enough time.

**Comment by dereuromark** [comment]
ping @dersam @sane4ek-2

**Comment by othercorey** [comment]
Closing due to no activity.

**Pull request closed without merging**